File: Agora-RTM-Tutorial-macOS-Objective-C/Agora-RTM-Tutorial/MessageCell.m
//
//  MessageCell.m
//  Agora-Rtm-Tutorial
//
//  Created by CavanSu on 2019/2/21.
//  Copyright © 2019 Agora. All rights reserved.
//

#import "MessageCell.h"

@interface MessageCell ()
@property (weak, nonatomic) IBOutlet NSTextField *leftUserLabel;
@property (weak, nonatomic) IBOutlet NSTextField *leftContentLabel;

@property (weak, nonatomic) IBOutlet NSTextField *rightUserLabel;
@property (weak, nonatomic) IBOutlet NSTextField *rightContentLabel;

@property (nonatomic, assign) CellType type;
@property (nonatomic, copy) NSString *user;
@property (nonatomic, copy) NSString *content;
@end

@implementation MessageCell

- (void)awakeFromNib {
    [super awakeFromNib];
    
    self.rightContentLabel.usesSingleLineMode = NO;
    self.rightContentLabel.cell.wraps = YES;
    self.rightContentLabel.cell.scrollable = NO;
    
    self.leftContentLabel.usesSingleLineMode = NO;
    self.leftContentLabel.cell.wraps = YES;
    self.leftContentLabel.cell.scrollable = NO;
}

- (void)setType:(CellType)type {
    _type = type;
    BOOL rightHidden = type == CellTypeLeft ? true : false;
    
    self.rightUserLabel.hidden = rightHidden;
    self.rightContentLabel.hidden = rightHidden;
    
    self.leftUserLabel.hidden = !rightHidden;
    self.leftContentLabel.hidden = !rightHidden;
}

- (void)setUser:(NSString *)user {
    _user = [user copy];
    switch (self.type) {
        case CellTypeLeft: self.leftUserLabel.stringValue = user; break;
        case CellTypeRight: self.rightUserLabel.stringValue = user; break;
    }
}

- (void)setContent:(NSString *)content {
    _content = [content copy];
    switch (self.type) {
        case CellTypeLeft: self.leftContentLabel.stringValue = content; break;
        case CellTypeRight: self.rightContentLabel.stringValue = content; break;
    }
}

- (void)updateType:(CellType)type message:(Message *)message {
    self.type = type;
    self.user = message.userId;
    self.content = message.text;
}

@end

File: Agora-Rtm-Tutorial-macOS/Agora-Rtm-Tutorial/MessageCell.swift
//
//  MessageCell.swift
//  Agora-Rtm-Tutorial
//
//  Created by CavanSu on 2019/1/17.
//  Copyright © 2019 Agora. All rights reserved.
//

import Cocoa

enum CellType {
    case left, right
}

class MessageCell: NSTableCellView {
    @IBOutlet weak var rightUserLabel: NSTextField!
    @IBOutlet weak var rightContentLabel: NSTextField!
   
    @IBOutlet weak var leftUserLabel: NSTextField!
    @IBOutlet weak var leftContentLabel: NSTextField!
    
    override func awakeFromNib() {
        super.awakeFromNib()
        #if os(iOS)
        rightUserBgView.layer.cornerRadius = 20
        rightContentBgView.layer.cornerRadius = 5
        
        leftUserBgView.layer.cornerRadius = 20
        leftContentBgView.layer.cornerRadius = 5
        #else
        rightUserLabel?.usesSingleLineMode = false
        rightUserLabel?.cell?.wraps = true
        rightUserLabel?.cell?.isScrollable = false
        
        leftContentLabel?.usesSingleLineMode = false
        leftContentLabel?.cell?.wraps = true
        leftContentLabel?.cell?.isScrollable = false
        #endif
    }
    
    private var type: CellType = .right {
        didSet {
            let rightHidden = type == .left ? true : false
            
            rightUserLabel.isHidden = rightHidden
            rightContentLabel.isHidden = rightHidden
        
            leftUserLabel.isHidden = !rightHidden
            leftContentLabel.isHidden = !rightHidden
            
            #if os(iOS)
            rightUserBgView.isHidden = rightHidden
            rightContentBgView.isHidden = rightHidden
            
            leftUserBgView.isHidden = !rightHidden
            leftContentBgView.isHidden = !rightHidden
            #endif
        }
    }
    
    private var user: String? {
        didSet {
            switch type {
            case .left:  leftUserLabel.stringValue = user!
            case .right: rightUserLabel.stringValue = user!
            }
        }
    }
    
    private var content: String? {
        didSet {
            switch type {
            case .left:  leftContentLabel.stringValue = content!
            case .right: rightContentLabel.stringValue = content!
            }
        }
    }
    
    func update(type: CellType, message: Message) {
        self.type = type
        self.user = message.userId
        self.content = message.text
    }
}
